### PR TITLE
Heading block: use toolbar for heading level control

### PR DESCRIPTION
**Pull request by ZebulanStanphill**
## Description
Having the heading level picker in both the toolbar and inspector is confusing and inconsistent. It's also weird that only the inspector version contains the h1, h5, and h6 options.

This PR removes the inspector control and puts all the heading level options into the toolbar dropdown.

## How has this been tested?
I opened up a post and played around with the heading level toolbar control to verify that it worked.

## Screenshots
image

**UPDATE**: based on feedback, I've updated the control to look like this:

image

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/heading/edit.js [hunk 1]
```diff
@@ -47,8 +45,8 @@ function HeadingEdit( {
 		<>
 			<BlockControls>
 				<HeadingToolbar
-					minLevel={ 2 }
-					maxLevel={ 5 }
+					minLevel={ 1 }
+					maxLevel={ 7 }
```
Comment (ZebulanStanphill) [inline packages/block-library/src/heading/edit.js]: No, this is not a mistake. The `maxLevel` is actually exclusive rather than inclusive, so you have to set it to 7 if you want to include heading level 6. I thought that was weird, but I decided it would be best to address it in a separate PR, if it needs to be addressed at all. Does anyone else think this is weird?

**Comment by paaljoachim** [comment]
Thanks for adjusting this feature Zeb!

**Comment by ZebulanStanphill** [comment]
The E2E test failure is because in `packages/e2e-tests/specs/editor/various/rich-text.test.js`, the `should handle change in tag name gracefully` test tries to click on one of the buttons in the inspector, which has been removed.

I'm not sure how to update the test to click the button in the block toolbar, however. Something I have just discovered is that the heading level button in the toolbar doesn't have any `aria-label`, so I can't click it from the E2E test the same way that the existing test works (by clicking a button with a matching `aria-label`).

I fixed the label issue in #20248. I guess that PR needs to be merged before this one so I can fix the E2E test.

**Comment by shaunandrews** [comment]
Would it be possible to do something like this, were we don't need to repetitive "Heading" labels for each option?

image

**Comment by ZebulanStanphill** [comment]
I've temporarily included changes from #20248 to show that, with the label fix from that PR merged and with the e2e test adjusted, all tests pass.

**Comment by ZebulanStanphill** [comment]
@shaunandrews I'm sure it would be possible, but it's out of the scope of this PR, and it might have negative effects on accessibility due to the removal of visible labels. I'd recommend opening a separate issue for that.

**Comment by karmatosed** [comment]
Whilst I am for them all going in toolbar, I do wonder if what @shaunandrews is suggesting would be a better visual. Whilst I understand the focus of this PR iterating on the design to be the best possible one I think might work here, or at least we should rapidly get an issue up for that.

**Comment by ZebulanStanphill** [comment]
#20248 has been merged into `master`, so I have rebased this branch, and it is now ready for review/merge!

File: packages/block-library/src/heading/edit.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-library/src/heading/edit.js]: The fact that heading 1 was omitted from the toolbar was a requirement/recommendation from the a11y team to discourage its usage as it's generally used for post title in the templates. Is this something we want to reconsider?

Comment (ZebulanStanphill) [reply packages/block-library/src/heading/edit.js]: @youknowriad Some kind of warning would be helpful, similar to the color contrast warning (which can be used in the toolbar, as demonstrated by the Navigation block). No equivalent currently exists for heading levels except in the Document Outline.

Showing a duplicate control in the inspector with the discouraged options is not the right answer, in my opinion. We should be dealing with incorrect heading levels the same way we deal with low-contrast background/text color combinations: by warning the user when they choose a potentially poor option.

**Comment by ZebulanStanphill** [comment]
Rebased; everything looks fine in the new G2 UI.

**Comment by karmatosed** [comment]
@ZebulanStanphill thanks for working on this. What I currently see is:

image

I still personally feel having just the idea that @shaunandrews shared would be better. Would you be able to work on that as an iteration?

**Comment by ZebulanStanphill** [comment]
@shaunandrews' mockup looks cool, but it seems like it might be less accessible than the existing dropdown. However, I'm willing to try it out.

Unfortunately, I'm not sure how to go about implementing it. There are no existing implementations of anything quite like his mockup. Do I use `ToolbarGroup` or `Toolbar`? `Dropdown` or `DropdownMenu`? It's not clear to me what the correct way to approach this is. If anyone has suggestions, please share them.

I also find it confusing that the toolbar has more visible information for each option when `isCollapsed` is enabled, compared to when it is disabled.

**Comment by ZebulanStanphill** [comment]
Well, speaking of dropdown menus, the newly-opened #20761 is quite relevant.

**Comment by ZebulanStanphill** [comment]
Rebased. It's worth noting that since #20745 was merged, (native) mobile already uses the toolbar exclusively for the heading level control.

**Comment by karmatosed** [comment]
@ZebulanStanphill this still seems to be using the full heading. Whilst I totally understand you are a little stuck on implementing that, I do still think it could be worth doing that over merge a half-point right now. @youknowriad would you be able to advise here to help get this back on track based on what Shaun suggested?

This is what I am seeing now:

image

**Comment by enriquesanchez** [comment]
I agree that the repeating 'Heading' label is not ideal here. As far as accessibility goes, there's two reasons why labels are important: 

1. Most of the time, icons alone are not enough to convey meaning. Labels are the primary and most reliable way of conveying meaning.

2. Users of speech recognition software (such as Dragon or Apple's Voice Control) need labels in order to identify the command they need to request when they want to interact with an element: "Click Publish".

While we have many buttons without labels in the editor, I think that in this particular case not having the label is not a a big issue. This is because the icon itself is a "label": H1, H2, etc.

**Comment by ZebulanStanphill** [comment]
It doesn't work quite right yet (the active option is not being highlighted, and the styling seems a bit off), and I might not be using the components properly, but I've updated the branch so now the dropdown looks like this:

image

Please review the code and point out any flaws/bugs with my approach! :slightly_smiling_face:

**Comment by jasmussen** [comment]
Thank you for this PR. I would tend to generally agree with this assessment.

> Having the heading level picker in both the toolbar and inspector is confusing and inconsistent. It's also weird that only the inspector version contains the h1, h5, and h6 options.

However Riad picks up on why things are the way they are in [this comment](https://github.com/WordPress/gutenberg/pull/20246#discussion_r381523459). An addition to that, though, was that previously the heading levels were not in a drop-down, so it was also prudent from a space point of view, to include only the most important ones. Given that it's now in a dropdown, that latter limitation is no longer there. 

So what it boils down to is the tension between heading level best practices, and usability — two aspects that commonly go hand in hand but in this situations are at odds. 

I would lean towards usability in this case, and include all heading levels. For a couple of reasons:

- We would still default to inserting an H2 heading level.
- We have a heading-level tool in the top toolbar that alerts you to incorrect heading levels.
- If you really want to break the levels, you can do that even today by selecting an H1 in the sidebar, it's just slightly less intuitive.
- Even if you don't pick an h1, you can still mess up your levels, as in jump from h2 to h4. 
- With full site editing, you might actually want an h1 heading, as it might indeed be the first heading on the page.

For that reason, I'd move forward with this one!

File: packages/block-library/src/heading/edit.js [hunk 2]
```diff
@@ -41,9 +39,9 @@ function HeadingEdit( { attributes, setAttributes, mergeBlocks, onReplace } ) {
 	return (
 		<>
 			<BlockControls>
-				<HeadingToolbar
-					minLevel={ Platform.OS === 'web' ? 2 : 1 }
-					maxLevel={ Platform.OS === 'web' ? 5 : 7 }
+				<HeadingLevelToolbar
+					minLevel={ 1 }
+					maxLevel={ 6 }
```
Comment (youknowriad) [inline packages/block-library/src/heading/edit.js]: Maybe we can drop the min/max prop all together?

File: packages/block-library/src/heading/wip-heading-level-toolbar.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+			icon={ <HeadingLevelIcon level={ selectedLevel } /> }
+			label={ __( 'Change heading level' ) }
+			popoverProps={ POPOVER_PROPS }
+		>
+			{ () => (
+				<ToolbarGroup
+					isCollapsed={ false }
+					icon={ <HeadingLevelIcon level={ selectedLevel } /> }
+					controls={ range( minLevel, maxLevel + 1 ).map( ( index ) =>
+						createLevelControl( index, selectedLevel, onChange )
```
Comment (youknowriad) [inline packages/block-library/src/heading/wip-heading-level-toolbar.js]: I don't find that the separate function brings too much value personally. An inline function might be clearer for the reader.

File: packages/block-library/src/heading/wip-heading-level-toolbar.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	} = props;
+
+	return (
+		<Toolbar
+			isCollapsed={ isCollapsed }
+			icon={ <HeadingLevelIcon level={ selectedLevel } /> }
+			label={ __( 'Change heading level' ) }
+			popoverProps={ POPOVER_PROPS }
+		>
+			{ () => (
```
Comment (youknowriad) [inline packages/block-library/src/heading/wip-heading-level-toolbar.js]: Does this component `Toolbar` support regular children, why a function?

**Comment by youknowriad** [comment]
@jasmussen @ZebulanStanphill The reasons the styles are not working properly is because with a Dropdown, the DOM element is not inserted inside its parent DOM element but it's inserted into a Portal.

This means, the styles that apply to the "block-contextual-toolbar" won't apply here. 

@jasmussen remember that discussion https://github.com/WordPress/gutenberg/pull/21064#discussion_r396321014 (it's the same issue), I guess it's time to make a generic fix (move the styles to `components-toolbar`)

File: packages/block-library/src/heading/wip-heading-level-toolbar.js (hunk 4 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/heading/wip-heading-level-toolbar.js]: I have no idea what exactly `Toolbar` supports, but when I tried nesting `ToolbarGroup` in it normally, the `Toolbar` popover was blank when opened. I saw the function pattern somewhere else, so I tried that and it worked. :man_shrugging:

File: packages/block-library/src/heading/edit.js (hunk 2 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/heading/edit.js]: Yeah, now that I think about it, there's no reason to keep the props if there's only one unified toolbar.

File: packages/block-library/src/heading/wip-heading-level-toolbar.js (hunk 4 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/heading/wip-heading-level-toolbar.js]: Actually, upon further inspection, I think I sort of get it now: when you don't pass an `__experimentalAccessibilityLabel` to `Toolbar`, `Toolbar` returns a `ToolbarGroup`, and if `isCollapsed` is passed to the `ToolbarGroup`, then it returns a `DropdownMenu`, and `DropdownMenu` takes a function.

**Comment by ZebulanStanphill** [comment]
I've simplified the code based on feedback. The active heading level option still isn't being styled properly due to the issues mentioned by @youknowriad. I'm still not sure if I'm using the components properly or if there's a better way to do this, but the toolbar does seem to be working as intended.

This is out-of-scope for this PR, but I think the freed-up space in the heading level toolbar could be used by an inline warning that appears when you select a questionable heading level, similar to the contrast checker in the Navigation block's color controls.

**Comment by ZebulanStanphill** [comment]
At least one of the recent test failures was legitimate: after switching to the icon-only toolbar in the popover, the same E2E test I updated before had to be updated again.

As for the other test failures, those appear to be gone thanks to the recent fixes in `master`. :smile:

**Comment by ZebulanStanphill** [comment]
Had to rebase again after #21039 was merged. :slightly_smiling_face:

**Comment by ZebulanStanphill** [comment]
After a rebase, the fixes from #21252 have improved the look of the heading level toolbar:
image

**Comment by shaunandrews** [comment]
Here's what I see currently in this PR:

Heading Options

I don't think we need the extra black border and the spacing it adds. Also, there seems to be some spacing issue when an option is both selected and focused — I think the focus state needs to be an outset border, to avoid that spacing issue.

Here's a quick update on the design I had in mind:

Heading Options Tweaks

**Comment by ZebulanStanphill** [comment]
@shaunandrews The focus state is consistent with other buttons, so I think changing that would be out-of-scope of this PR. I think the issue might be the size of the buttons themselves compared to the ones in the main block toolbar, though again, the styles are inherited from elsewhere and changing them should probably happen in another PR.

As for the black border, I'm not sure how to go about getting rid of it in this context; I'm still not even sure I'm using all the Toolbar/Popover components properly. Right now I'm using an `isCollapsed: false` `ToolbarGroup` inside an `isCollpased: true` `Toolbar`, but that doesn't quite feel right to me.

What should I be targeting to remove the border? `ToolbarGroup`s nested inside `Popover`s? In exactly what contexts should a `ToolbarGroup` have or not have its black border?

Also, if you view the heading level popover as possibly containing more than just the toolbar (e.g. an inline warning for invalid heading level nesting), the toolbar group border starts to make more sense.

**Comment by ZebulanStanphill** [comment]
After a rebase including the changes from #21476, the toolbar now looked like this:

image

I then added some styles to the Heading block `editor.scss` file in d16da4e to make the popover look nicer:

image

However, I'm not satisfied with the implementation of these styles. They feel too specific to this particular block. I could use some help figuring out a better way to implement this.

It's worth noting that this popover will likely also be used to hold an inline warning for invalid heading levels in the future, so whatever styles are used have to account for that being present.

File: packages/block-library/src/heading/heading-level-toolbar.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+export default function HeadingLevelToolbar( { selectedLevel, onChange } ) {
+	return (
+		<Toolbar
+			isCollapsed={ true }
+			icon={ <HeadingLevelIcon level={ selectedLevel } /> }
+			label={ __( 'Change heading level' ) }
+			popoverProps={ POPOVER_PROPS }
+		>
+			{ () => (
```
Comment (youknowriad) [inline packages/block-library/src/heading/heading-level-toolbar.js:45]: @gziolo @diegohaz is this the correct way to use Toolbar and ToolbarGroup?

**Review by youknowriad — COMMENTED** [review]
Looks good to me code wise, let's get a design review and ship

File: packages/block-library/src/heading/heading-level-toolbar.js (hunk 5 above)
Comment (diegohaz) [reply packages/block-library/src/heading/heading-level-toolbar.js:45]: The wrapper in this case should be a `ToolbarGroup` because it’s meant to be contained in a `role=“toolbar” element (the block toolbar).

From my understanding, the `ToolbarGroup` within the popover should be a sort of `DropdownMenu` instead. But it’s confusing to me that it’s horizontal, and not vertical as the other toolbar controls.

I see that it was designed vertically at first. I’m not sure how I feel with it looking similar to the control in the inspector, and not to the other toolbar controls.

Comment (ZebulanStanphill) [reply packages/block-library/src/heading/heading-level-toolbar.js:45]: @diegohaz I used `Toolbar` because that's what `BlockAlignmentToolbar` uses. Notably, `Toolbar` basically just acts as a wrapper for `ToolbarGroup` anyway:

```js
function Toolbar( { className, __experimentalAccessibilityLabel, ...props } ) {
	if ( __experimentalAccessibilityLabel ) {
		return (
			<ToolbarContainer
				// `ToolbarGroup` already uses components-toolbar for compatibility reasons
				className={ classnames(
					'components-accessible-toolbar',
					className
				) }
				accessibilityLabel={ __experimentalAccessibilityLabel }
				{ ...props }
			/>
		);
	}

	return <ToolbarGroup { ...props } className={ className } />;
}
```

`ToolbarContainer` contains a nested `Toolbar` which end up rendering a `ToolbarGroup`.

So is `BlockAlignmentToolbar` doing it wrong, and should I just use `ToolbarGroup` directly?

Also, with regard to using `DropdownMenu`, it's worth noting that I think an inline warning for invalid heading levels should appear in the dropdown in the future. Does it still make sense to use `DropdownMenu` in that context? (It's also worth noting that my `isCollapsed` `Toolbar` is already using `Dropdown`.

The relation between all these different components is really confusing. :confounded:

Comment (diegohaz) [reply packages/block-library/src/heading/heading-level-toolbar.js:45]: Once `__experimentalAccessibilityLabel` is stable, using `Toolbar` as a `ToolbarGroup` (that is, without passing an `accessibilityLabel` prop) will be deprecated.

And, conceptually speaking, having nested toolbars (`role="toolbar"`) doesn't make sense.

`BlockAlignmentToolbar` hasn't been migrated yet. It's not a big deal though as the migration is simple (just renaming `Toolbar` -> `ToolbarGroup`).

I totally agree that it's confusing, especially when you take into consideration the various possibilities of popovers (`Popover`, `Dropdown`, `DropdownMenu`, `isCollapsed`, `controls` or `children`).

Comment (ZebulanStanphill) [reply packages/block-library/src/heading/heading-level-toolbar.js]: The whole `isCollapsed` thing is especially confusing. It takes a toolbar group and then creates a popover around it with a toggle to open the popover, while also changing the appearance of the toolbar to use visible text labels for every option in the toolbar. If a group of controls is hidden in a popover like that, are they still part of the parent toolbar (and therefore a "toolbar group"), or are they actually a separate toolbar (something that should have its own role) accessed by pressing a button in another toolbar?

I get the feeling that the `isCollapsed` prop was a questionable design choice since whether or not it's toggled drastically alters what kind of component is actually being rendered.

There's also the aforementioned idea of putting an inline heading level warning in the popover (similar to the color contrast warning for color controls). At that point, is the heading level thing still just a toolbar group, or is it a whole panel?

Anyway, I've pushed some changes to the implementation based on your feedback. Let me know what you think.

**Comment by shaunandrews** [comment]
I noticed something strange happening with the tooltips:

something funky

The H1 tooltip appears when I open the menu for some reason, it then repositions itself and sticks around around I hover over/out of the H1 option.

**Comment by ZebulanStanphill** [comment]
@shaunandrews If I had to guess, it probably has something to do with the toolbar button being focused (and thus causing the tooltip to appear) before its open animation is finished. There have been several issues like that with popovers and their positions not being adjusted quickly enough. Whatever the case, fixing that is out-of-scope for this PR.

**Review by shaunandrews — APPROVED** [review]

**Comment by youknowriad** [comment]
Here's what I see when 

Capture d’écran 2020-05-08 à 3 22 38 PM

Isn't that too much borders?

**Comment by ZebulanStanphill** [comment]
@youknowriad Yeah, it's not supposed to look like that. The styles in `editor.scss` are not targeting the right elements at the momet. I'm in the middle of refactoring the PR a bit, and gutenberg.run doesn't seem to be working at the moment (and `wp-env` was half-broken on Linux last time I tried to use it), so testing it is taking a long time. :disappointed: I'm trying to see if I need to change anything after #20008.

**Comment by SergioEstevao** [comment]
@ZebulanStanphill Just a heads up that you would like to test this on GB-Mobile before this is merged. We share the same block code, so it will be good that the code works on GB-mobile too.

**Review by SergioEstevao — CHANGES_REQUESTED** [review]
Please wait until this is tested on GB-mobile.

**Comment by ZebulanStanphill** [comment]
@SergioEstevao I'll let you know when the PR is ready for reviews again. :+1:

**Comment by ZebulanStanphill** [comment]
Alright, I've now got this PR working with the changes from #20008, so the Heading block toolbar will remain one tab stop, rather than falling back to multiple-tab-stops.

The implementation has changed quite a bit, and I'm still not sure if I'm using the components correctly. I'd definitely appreciate technical and accessibility reviews. The way the heading level control is presented in this PR is currently unique: I'm using a whole `Toolbar` (not just `ToolbarGroup`) inside the popover. (Keep in mind I'm thinking that there will eventually also be an inline warning in the popover for using invalid combinations of heading levels.) Is this the correct approach?

Also, I made the popover the exact width of the toolbar:

image

Some end-to-end tests are currently failing, but they look unrelated, and there are failing tests on other PRs right now, so I assume the fault is with the tests, not the PR. 

@SergioEstevao This PR should now be ready for someone to test with Gutenberg Mobile.

**Comment by SergioEstevao** [comment]
As I feared this solution doesn't work well for the Mobile version. 

Screenshot 2020-05-11 at 15 06 10

For the RN mobile version, I would like to keep the old dropdown approach, to make it consistent with other options like alignment. It will look like this:

Screenshot 2020-05-11 at 15 11 41
 

@ZebulanStanphill do you mind I push some commits to your branch for a fix for that?

**Comment by ZebulanStanphill** [comment]
@SergioEstevao Feel free to push as many commits as necessary. :slightly_smiling_face::+1:

**Comment by SergioEstevao** [comment]
> In case anyone is wondering, I put the ToolbarGroup outside of the HeadingLevelDropdown component because I figured that long-term, the text alignment and heading level buttons should be in the same toolbar group (meaning no dividing line between them). Changing the text alignment toolbar dropdown would have been outside the scope of this PR, so I never included it.

> Personally, I think the whole isCollapsed thing stopped making sense when the controls in the block toolbars became always-collapsed. The isCollapsed prop suddenly turns a toolbar group into a button opening a dropdown containing a different style of toolbar. That's why I went with a component name of HeadingLevelDropdown rather than HeadingLevelToolbar.

@ZebulanStanphill I agree with your comment above so I updated my code to move the ToolbarGroup to outside the heading-level-dropdown. 

Can you give it another look?

**Review by SergioEstevao — APPROVED** [review]
This is now ok from a  GB-mobile perspective.

**Comment by youknowriad** [comment]
Testing on safari, I have a scrollbar

Capture d’écran 2020-05-21 à 11 16 20 AM

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+											onClick() {
+												onChange( targetLevel );
+											},
+										};
+									}
+								) }
+							/>
+						</Toolbar>
+					) }
+				></Dropdown>
```
Comment (youknowriad) [inline packages/block-library/src/heading/heading-level-dropdown.js:111]: Can we just use a self-closing component here?

File: packages/block-library/src/heading/heading-level-icon.js [hunk 7]
```diff
@@ -3,6 +3,24 @@
...
 
+/** @typedef {import('@wordpress/element').WPComponent} WPComponent */
+
+/**
+ * HeadingLevelIcon props.
+ *
+ * @typedef WPHeadingLevelIconProps
+ *
+ * @property {number}   level     The heading level to show an icon for.
+ * @property {?boolean} isPressed Whether or not the icon should appear pressed; default: false.
```
Comment (youknowriad) [inline packages/block-library/src/heading/heading-level-icon.js:14]: is this how we mark optional types, I thought it was  more `boolean=`

**Review by youknowriad — APPROVED** [review]
This is looking very good for me. Feel free to ship once the scrollbar is fixed.

File: packages/block-library/src/heading/heading-level-icon.js (hunk 7 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/heading/heading-level-icon.js:14]: I'm using the `?xxx` format because [these docs](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#nullable-undefined-and-void-types) said to use it for nullable types.

**Comment by ZebulanStanphill** [comment]
The Safari scrollbar issue has something to do with `min-width: max-content` resulting in a smaller value on Safari than on Chromium. I don't know why. For now, I'm just going to go with a hardcoded `230px`, with a TODO note to try and find a better way to style it in the future.

Also, could you test one more time to make sure the toolbar is actually usable on Safari? I think I might have had something weird going on on my end, but when I tried to use the heading level toolbar on that browser, I couldn't click the buttons - it would just close the toolbar with no effect. Using the keyboard worked, though.

**Comment by youknowriad** [comment]
I  have the same issue, not being able to click on buttons.

**Comment by ZebulanStanphill** [comment]
Well, I have no idea why it's happening. :disappointed: Any ideas?

**Comment by youknowriad** [comment]
It's probably related to the difference between "focus on click" behavior between  browsers. Some browsers like Chrome do focus the button (so it avoids closing the popover) and others like Safari don't focus elements on click which doesn't impact  in general but I suspect that the focus is here moved elsewhere by some other code (RichText or the block for  some reason) which closes the popover before the click handler happens?

**Comment by youknowriad** [comment]
checking what element is focused (activeElement) when the closing happens can give you an idea about why exactly it's happening.

**Comment by paaljoachim** [comment]
I checked out the PR at gutenberg.run.
It works fine in Chrome, but as you both mention, in Safari the buttons do not react to the changing of heading levels.

Here is an animated gif I made from using Chrome.
New-Heading-block

Will it give a warning when selecting Heading level 1? As a person might select it because it has the biggest font size without knowing better.

**Comment by ZebulanStanphill** [comment]
@paaljoachim Well, in a future PR, I was planning to add an inline warning to the dropdown when you chose a questionable heading choice, similar to the color contrast checker that color controls have.

Another idea I just had is that the H1 option could be hidden when the Heading block detects it is in a `post_content`, which I think would be possible using block context. But again, I'd do that in another PR.

**Comment by ZebulanStanphill** [comment]
@youknowriad I checked; after clicking a button in the heading level toolbar, the `document.activeElement` is the document `<body>`.

**Comment by youknowriad** [comment]
@ZebulanStanphill I guess this means there's a rerender somewhere causing a "remount" which also means "focus loss".

**Comment by ZebulanStanphill** [comment]
As far as I can tell, I'm not doing anything to cause a remount, so I assume there must be a pre-existing bug somewhere?

Anyway, I've rebased the PR, squashed a few commits, and updated the dropdown/toolbar class names to better follow our naming standards.

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+				<Dropdown
+					popoverProps={ POPOVER_PROPS }
+					renderToggle={ ( { onToggle, isOpen } ) => {
+						const openOnArrowDown = ( event ) => {
+							if ( ! isOpen && event.keyCode === DOWN ) {
+								event.preventDefault();
+								event.stopPropagation();
+								onToggle();
+							}
+						};
```
Comment (diegohaz) [inline packages/block-library/src/heading/heading-level-dropdown.js:58]: I noticed that pressing arrow keys to move between the block toolbar buttons isn't working when this dropdown toggle button has focus, and I guess that's because this `openOnArrowDown` is overwriting `itemProps.onKeyDown`.

To fix that, you'll have to call `itemProps.onKeyDown` here like the suggestion below. But I wonder if just using `ToolbarButton` instead of `ToolbarItem` + `Button` wouldn't be better.
```suggestion
						const openOnArrowDown = ( event ) => {
							if ( itemProps.onKeyDown ) {
						        	itemProps.onKeyDown( event );
							}
							if ( ! isOpen && event.keyCode === DOWN ) {
								event.preventDefault();
								event.stopPropagation();
								onToggle();
							}
						};
```

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+													level={ targetLevel }
+													isPressed={ isActive }
+												/>
+											),
+											title: sprintf(
+												// translators: %s: heading level e.g: "1", "2", "3"
+												__( 'Heading %d' ),
+												targetLevel
+											),
+											isActive,
```
Comment (diegohaz) [inline packages/block-library/src/heading/heading-level-dropdown.js:101]: Here's a workaround to solve the problem on Safari and Firefox:
```suggestion
											isActive,
											onMouseDown( event ) {
												event.preventDefault();
												event.currentTarget.focus();
											}
```

**Review by diegohaz — COMMENTED** [review]
Great work @ZebulanStanphill!

I tested it on Chrome, Safari and Firefox and can confirm that the issue with closing the popover when pressing any of the heading toolbar buttons happens on Safari and Firefox (on MacOS). So, it's probably related to [buttons not receiving focus on click on those browsers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus). I just suggested a temporary workaround for that. But I'll try to find a better solution in the Toolbar module itself.

Also, arrow keys aren't working on the heading dropdown toggle button. And a simple solution for that is to just use `ToolbarButton` instead of `ToolbarItem`.

**Comment by ZebulanStanphill** [comment]
Thanks for the help, @diegohaz! I've switched to using a standard `ToolbarButton` and I've applied the macOS Firefox/Safari fix you suggested (and tested it on a friend's computer to confirm it worked).

Everything appears to be working as intended now! :smile:

**Review by diegohaz — APPROVED** [review]
Tested it on Chrome, Firefox and Safari, and it looks great! Also, I was able to use it with VoiceOver + Safari and NVDA + Firefox without problems. Couldn't find any other issue.

The only thing that feels odd to me is that the initial focus isn't on the current selected item. But this is probably something out of the scope of this PR as the same happens to the text alignment option.

**Comment by paaljoachim** [comment]
I now also works in Safari. Good job Zeb!

**Comment by paaljoachim** [comment]
One thing though.
I noticed some jumping of the black inserter during selection of various Heading levels when checking Safari. As I am using Gutenberg.run to check. At the moment I am having difficulty checking again. I was going to see if the same effect was in Chrome and Firefox as well. Changing the size of the heading should not as I see it make the inserter at all move.

**Comment by ZebulanStanphill** [comment]
>Changing the size of the heading should not as I see it make the inserter at all move.

I tried it out myself, and the behavior appears totally normal to me. Each heading size is a different height, so of course the appender would move up/down. The block is changing height, so the appender moves to stay below the block. Or are you talking about something else?

**Comment by paaljoachim** [comment]
Hey Zeb. Yes, I am talking about the moving inserter like so:

Heading-Inserter-jumps

But I do understand what your saying. As the size of the block changes the location of the inserter changes. Making it jump up or down based on which heading size is selected.
I just wanted to mention it.
Alrighty moving on..:)
Looks like the Heading block is ready for merge..:)

**Review by youknowriad — APPROVED** [review]
Nice work here. Thanks for being perseverant with the PR.

**Comment by ZebulanStanphill** [comment]
Now that this PR is out of the way, I've started working on adding a heading level checker to the block to encourage more semantic heading level choices: #22650.
